From 9a0aef07ba507e27300a80ab8c772832f1c8e0e2 Mon Sep 17 00:00:00 2001
From: Balamurugan Selvarajan <quic_bselvara@quicinc.com>
Date: Thu, 12 May 2022 16:52:00 +0530
Subject: [PATCH] ath12k: Add pageable mem type for QCN9274

Added pageable mem type region in the fixed mem.
This would be used for for dynamic Paging arena for QCN9274.

Signed-off-by: Balamurugan Selvarajan <quic_bselvara@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h     |  2 +
 drivers/net/wireless/ath/ath12k/coredump.c | 19 +++++
 drivers/net/wireless/ath/ath12k/qmi.c      | 96 +++++++++++++---------
 drivers/net/wireless/ath/ath12k/qmi.h      |  1 +
 4 files changed, 78 insertions(+), 40 deletions(-)

Index: backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/core.h
===================================================================
--- backports-20220404-5.4.164-f40abb4788.orig/drivers/net/wireless/ath/ath12k/core.h
+++ backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/core.h
@@ -924,6 +924,7 @@ struct ath12k_base {
 	u64 fw_soc_drop_count;
 
 	struct device_node *hremote_node;
+	u32 host_ddr_fixed_mem_off;
 
 	/* must be last */
 	u8 drv_priv[0] __aligned(sizeof(void *));
@@ -1075,6 +1076,9 @@ void ath12k_core_halt(struct ath12k *ar)
 int ath12k_core_resume(struct ath12k_base *ab);
 int ath12k_core_suspend(struct ath12k_base *ab);
 int ath12k_config_qdss(struct ath12k_base *ab);
+int ath12k_coredump_mhi_update_bhie_table(struct ath12k_base *ab,
+					  void __iomem *va,
+					  phys_addr_t pa, size_t size);
 
 const struct firmware *ath12k_core_firmware_request(struct ath12k_base *ab,
 						    const char *filename);
Index: backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/coredump.c
===================================================================
--- backports-20220404-5.4.164-f40abb4788.orig/drivers/net/wireless/ath/ath12k/coredump.c
+++ backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/coredump.c
@@ -311,3 +311,21 @@ out:
 	kfree(dump);
 }
 
+int ath12k_coredump_mhi_update_bhie_table(struct ath12k_base *ab,
+					  void __iomem *va,
+					  phys_addr_t pa, size_t size)
+{
+	struct ath12k_pci *ar_pci = (struct ath12k_pci *)ab->drv_priv;
+	struct mhi_controller *mhi_ctrl = ar_pci->mhi_ctrl;
+	int ret;
+
+	/* Attach Pageable region to MHI buffer so that it is
+	 * included as part of pageable region in dumps
+	 */
+	ret = mhi_update_bhie_table_for_dyn_paging(mhi_ctrl, va, pa, size);
+	if (ret)
+	        ath12k_dbg(ab, ATH12K_DBG_QMI,
+			   "failed to add Dynamic Paging region to MHI Buffer table %d\n", ret);
+
+	return ret;
+}
Index: backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/qmi.c
===================================================================
--- backports-20220404-5.4.164-f40abb4788.orig/drivers/net/wireless/ath/ath12k/qmi.c
+++ backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/qmi.c
@@ -2433,70 +2433,89 @@ static int ath12k_qmi_assign_target_mem_
 	struct device *dev = ab->dev;
 	struct device_node *hremote_node = NULL;
 	struct resource res;
-	u32 host_ddr_sz;
+	int host_ddr_sz, sz;
 	int i, idx, ret;
 
-	for (i = 0, idx = 0; i < ab->qmi.mem_seg_count; i++) {
-		switch (ab->qmi.target_mem[i].type) {
-		case HOST_DDR_REGION_TYPE:
-			hremote_node = of_parse_phandle(dev->of_node, "memory-region", 0);
-			if (!hremote_node) {
-				ath12k_warn(ab, "qmi fail to get hremote_node\n");
-				return ret;
-			}
+	sz = ab->host_ddr_fixed_mem_off;
+	hremote_node = of_parse_phandle(dev->of_node, "memory-region", 0);
+	if (!hremote_node) {
+		ath12k_warn(ab, "qmi fail to get hremote_node\n");
+		return ret;
+	}
 
-			ret = of_address_to_resource(hremote_node, 0, &res);
-			if (ret) {
-				ath12k_warn(ab, "qmi fail to get reg from hremote\n");
-				return ret;
-			}
+	ret = of_address_to_resource(hremote_node, 0, &res);
+	if (ret) {
+		ath12k_warn(ab, "qmi fail to get reg from hremote\n");
+		return ret;
+	}
 
-			if (res.end - res.start + 1 < ab->qmi.target_mem[i].size) {
-				ath12k_warn(ab, "qmi fail to assign memory of sz\n");
-				return -EINVAL;
-			}
+	host_ddr_sz = (res.end - res.start) + 1;
 
-			ab->qmi.target_mem[idx].paddr = res.start;
+	for (i = 0, idx = 0; i < ab->qmi.mem_seg_count; i++) {
+		ab->qmi.target_mem[idx].paddr = res.start + sz;
+
+		if (ab->qmi.target_mem[i].size > (host_ddr_sz - sz)) {
+			ath12k_warn(ab, "No fixed mem to assign for type %d\n",
+				    ab->qmi.target_mem[i].type);
+			return -EINVAL;
+		}
+
+		switch (ab->qmi.target_mem[i].type) {
+		case HOST_DDR_REGION_TYPE:
 			ab->qmi.target_mem[idx].v.ioaddr =
 					ioremap(ab->qmi.target_mem[idx].paddr,
 						ab->qmi.target_mem[i].size);
-			ab->qmi.target_mem[idx].size = ab->qmi.target_mem[i].size;
-			host_ddr_sz = ab->qmi.target_mem[i].size;
-			ab->qmi.target_mem[idx].type = ab->qmi.target_mem[i].type;
+			ab->qmi.target_mem[idx].size =
+					ab->qmi.target_mem[i].size;
+			ab->qmi.target_mem[idx].type =
+					ab->qmi.target_mem[i].type;
+			sz += ab->qmi.target_mem[i].size;
 			idx++;
 			break;
 		case CALDB_MEM_REGION_TYPE:
-			if (ab->qmi.target_mem[i].size > ATH12K_QMI_CALDB_SIZE) {
-				ath12k_warn(ab, "qmi mem size is low to load caldata\n");
-				return -EINVAL;
-			}
-
-			if (ath12k_cold_boot_cal && ab->hw_params.cold_boot_calib) {
-				if (hremote_node) {
-					ab->qmi.target_mem[idx].paddr =
-							res.start + host_ddr_sz;
-					ab->qmi.target_mem[idx].v.ioaddr =
-						ioremap(ab->qmi.target_mem[idx].paddr,
-							ab->qmi.target_mem[i].size);
-				} else {
-					ab->qmi.target_mem[idx].paddr =
-						ATH12K_QMI_CALDB_ADDRESS;
-				}
+			if (ath12k_cold_boot_cal &&
+				ab->hw_params.cold_boot_calib) {
+				ab->qmi.target_mem[idx].v.ioaddr =
+					ioremap(ab->qmi.target_mem[idx].paddr,
+						ab->qmi.target_mem[i].size);
+				sz += ab->qmi.target_mem[i].size;
 			} else {
 				ab->qmi.target_mem[idx].paddr = 0;
 				ab->qmi.target_mem[idx].v.ioaddr = NULL;
 			}
 
-			ab->qmi.target_mem[idx].size = ab->qmi.target_mem[i].size;
-			ab->qmi.target_mem[idx].type = ab->qmi.target_mem[i].type;
+			ab->qmi.target_mem[idx].size =
+					ab->qmi.target_mem[i].size;
+			ab->qmi.target_mem[idx].type =
+					ab->qmi.target_mem[i].type;
 			idx++;
 			break;
+		case PAGEABLE_MEM_TYPE:
+                        ab->qmi.target_mem[idx].v.ioaddr =
+                                        ioremap(ab->qmi.target_mem[idx].paddr,
+						ab->qmi.target_mem[i].size);
+                        ab->qmi.target_mem[idx].size =
+					ab->qmi.target_mem[i].size;
+                        ab->qmi.target_mem[idx].type =
+					ab->qmi.target_mem[i].type;
+			ret = ath12k_coredump_mhi_update_bhie_table(ab,
+					ab->qmi.target_mem[idx].v.ioaddr,
+					ab->qmi.target_mem[idx].paddr,
+					ab->qmi.target_mem[idx].size);
+			sz += ab->qmi.target_mem[i].size;
+			if (ret < 0)
+				ath12k_warn(ab,
+				            "qmi fail to update BHI table %d\n", ret);
+			idx++;
+			break;
+
 		default:
 			ath12k_warn(ab, "qmi ignore invalid mem req type %d\n",
 				    ab->qmi.target_mem[i].type);
 			break;
 		}
 	}
+	ab->host_ddr_fixed_mem_off = sz;
 	ab->qmi.mem_seg_count = idx;
 
 	return 0;
@@ -3250,7 +3269,14 @@ int ath12k_qmi_pci_alloc_qdss_mem(struct
 			return ret;
 		}
 
-		ab->qmi.qdss_mem[0].paddr = (phys_addr_t)res.start + QMI_Q6_QDSS_ETR_OFFSET_QCN92XX;
+		if ((ab->host_ddr_fixed_mem_off + ab->qmi.qdss_mem[0].size) >
+			((res.end - res.start) + 1)) {
+			ath12k_warn(ab,
+				    "No Fixed mem to allocate for QDSS_ETR_MEM_REGION_TYPE");
+			return -EINVAL;
+		}
+
+		ab->qmi.qdss_mem[0].paddr = (phys_addr_t)res.start + ab->host_ddr_fixed_mem_off;
 		ab->qmi.qdss_mem[0].v.ioaddr =
 			ioremap(ab->qmi.qdss_mem[0].paddr,
 				ab->qmi.qdss_mem[0].size);
@@ -3258,6 +3284,7 @@ int ath12k_qmi_pci_alloc_qdss_mem(struct
 			ath12k_warn(ab, "WARNING etr-addr remap failed\n");
 			return -ENOMEM;
 		}
+		ab->host_ddr_fixed_mem_off += ab->qmi.qdss_mem[0].size;
 		break;
 	default:
 		ath12k_warn(ab, "qmi ignore invalid qdss mem req type %d\n",
Index: backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/qmi.h
===================================================================
--- backports-20220404-5.4.164-f40abb4788.orig/drivers/net/wireless/ath/ath12k/qmi.h
+++ backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/qmi.h
@@ -211,6 +211,7 @@ struct qmi_wlanfw_qdss_trace_mode_resp_m
 #define BDF_MEM_REGION_TYPE				0x2
 #define M3_DUMP_REGION_TYPE				0x3
 #define CALDB_MEM_REGION_TYPE				0x4
+#define PAGEABLE_MEM_TYPE				0x9
 #define QMI_WLANFW_MAX_PLATFORM_NAME_LEN_V01		64
 #define QMI_WLANFW_MAX_HOST_DDR_RANGE_SIZE_V01		3
 #define QDSS_ETR_MEM_REGION_TYPE                        0x6
Index: backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/core.c
===================================================================
--- backports-20220404-5.4.164-f40abb4788.orig/drivers/net/wireless/ath/ath12k/core.c
+++ backports-20220404-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/core.c
@@ -923,6 +923,8 @@ static void ath12k_core_reset(struct wor
 	ath12k_dbg(ab, ATH12K_DBG_BOOT, "reset starting\n");
 
 	ab->is_reset = true;
+	/* reset host fixed mem off to zero */
+	ab->host_ddr_fixed_mem_off = 0;
 	/* prepare coredump */
 	ath12k_coredump_download_rddm(ab);
 
